hypothesis-python/docs/explanation/shrinking.rst

```diff
@@ -0,0 +1,138 @@
+
+    @given(st.lists(st.integers()))
+    @settings(report_multiple_bugs=False, database=None)
+    def f(lst):
+        print(lst, "<-- failure" if len(lst) > 1 else "")
+        assert len(lst) <= 1
+    f()
+
+Here's my output, where we can see this pass removing list elements which are not relevant to the failure:
+
+.. code-block:: none
+
+    []
+    [0]
+    [19578, -3592, 4925] <-- failure
+    [19578, -3592, 4925] <-- failure
+    [19578, -3592, 0] <-- failure
+    [19578, -3592] <-- failure
+    [19578, 0] <-- failure
+    [19578]
+    [0, 0] <-- failure
+    [0, 0] <-- failure
+
+Hypothesis generates ``[19578, -3592, 4925]`` as the first failure, then transitions into shrinking it. It tries removing elements in succession, then realizes that removing any further elements will not be useful once it hits ``[19578]``, which is not a failure. It then reduces each list element and ends at the minimal failing example of ``[0, 0]``.
+
+Other shrinking passes
+----------------------
+
+These two passes (reducing individual choices and removing choices) form the core of the shrinker. However, there are a number of additional shrinking passes in Hypothesis. Most of these reduce failures that have a specific relationship between two or more choices. For instance, one shrinking pass tries to balance two integers so they add up to some target (which the shrinker doesn't know about ahead of time):
+
+.. code-block:: python
+
+    from hypothesis import given, settings, strategies as st
+
+    @given(st.integers(), st.integers())
+    @settings(report_multiple_bugs=False, database=None)
+    def f(n1, n2):
+        print(n1, n2, "<-- failure" if n1 >= 5 and n1 + n2 >= 50 else "")
+        if n1 >= 5:
+            assert n1 + n2 < 50
+    f()
+
+Here's a portion of the output, focusing on when this shrinking pass gets ran:
+
+.. code-block:: none
+
+    ...
+    -8 41
+    8 42 <-- failure
+    7 43 <-- failure
+    6 44 <-- failure
+    5 45 <-- failure
+    4 46
+    4 44
+    ...
+
+We can see that this pass increases ``n1`` by the same amount that it decreases ``n2`` by. Hypothesis eventually reports that the failure ``n1=5, n2=45``, which is in fact the minimal failing example.
+
+Definition of example complexity
+--------------------------------
+
+How does the shrinker know when one example is "simpler" than another? Hypothesis defines a total ordering over the complexity of examples by using the :doc:`choice sequence <choice-sequence>`. Examples are ordered first by the number of choices. If one example makes fewer choices than another, the shrinker will consider it to be simpler, regardless of the type of those choices. If two examples make the same number of choices, the shrinker then orders them by a type-specific notion of complexity for each of the five choice sequence types. For example, the shrinker orders integers as ``0, 1, -1, 2, -2, 3, -3, ...``, with ``0`` being the simplest integer choice.
```

Comment:
One thing that IMO would be good to reinforce, either here or in the section on strategies, is that it is the *choice sequence* that is shrunk, not the *value*. For example, 
```
@given(st.integers() | st.just(2**32-1))
def test_it(i):
    assert not 2**16 <= i < 2**32
```
would not usually shrink the value from 2**32-1 to 2**16, since the value of the failing example would not carry over to the preferred `integers()` strategy. Related to the un-shrinkability of explicit examples, I guess.

I don't know if this is a common misconception to have, but at least to me this took some time to internalize.

Reply:
this is an extremely common misconception and worth stating explicitly!